### PR TITLE
Consistent use of singleton types in manual.

### LINKED ISSUE #7135: Manual could be clearer about parametric singleton types
I was confused about the meaning of parametric singleton types such as `Type{Array{T}}`, and the manual section "Singleton Types" does not say anything about them.  It should be improved.

The only documentation section I could find that says anything about parametric singleton types is "Case Study: Rational Conversions", which is misleading because it defines methods for `convert{T<:Int}(::Type{Rational{T}}, x::...)` but not  `convert(::Type{Rational}, x::...)`.  At first glance it seems like `convert(Rational, x)` would invoke one of those `Type{Rational{T}}` methods just as `convert(Rational{Int8}, x)` would.  In fact `convert(Rational, x)` invokes a method which is not in this documentation section.  The convert methods for rational numbers in base/rational.jl are somewhat different from what the documentation says.

Here's a simple example: define `f{T}(::Type{Array{T}}) = 1` then `f(Array{Int})` invokes that method but `f(Array)` and `f(Array{Int,1})` do not.  This makes perfect sense when you think about it, but it isn't really obvious and some specific documentation would help.

BTW unlike the 0.2 version, in the latest version manual on docs.julialang.org the section "Singleton Types" does not show up in the table of contents at all.  I am not sure what the problem is.  It can be found by going to "Parametric Types" in the table of contents and then scrolling down a long way.

### PR DESCRIPTION
**Pull request by tpapp**
Fixes #7135, see discussion there.

### Changes

- separate discussion of singleton types (as in `Base.issingletontype`) and `Type{T}`

- move discussion about singleton types to a later section so that parametric singleton types can be discussed

- add example and discussion about parametric singleton types

- add usage hint about `Type` outside dispatch ~~(TL;DR: don't)~~ example: type sharpening (thanks @tkf)

### Open questions:

- [x] is there a terminology for `Type{T}` now that we don't confuse it with singleton types? If not, do we want to introduce one, eg call it "type selectors"? 

- [x] should I mention that abstract types can't be singleton types, or is it obvious from the context?

- [x] should I expunge the remaining use of 'singleton' for `Type{T}` from *Conversion and Promotion*?

### CODE REVIEW

**Comment by tkf** [comment]
> add usage hint about `Type` outside dispatch (TL;DR: don't)

A usage (workaround) outside dispatch is to get a sharp field type.  For example, the type of `Fix1(Vector, undef)` is `Fix1{Type{Vector},UndefInitializer}` rather than `Fix1{DataType,UndefInitializer}`.  This is required for the field type to be inferable.  So, I think it'd be better to avoid stressing that `Type{T}` is _only_ for dispatch.  "`Type{T}` types" section in this PR seems to be fine as-is, though.

File: doc/src/manual/types.md
Comment (tkf) [inline doc/src/manual/types.md:1064]: I think the spirit of this formalism is great, but if I have

```julia
julia> abstract type A end

julia> struct B <: A end
```

and I don't define any other subtype of `A`, would it imply that `A` is also a singleton type?

Maybe making this perfect requires `isconcretetype(T)`?  Although if you say "for any `a` and `b`" I guess it could mean "for any `a` and `b` of any implementable type."

> should I mention that abstract types can't be singleton types, or is it obvious from the context?

So yes, I think this helps to make this definition clearer.

**Comment by tkf** [comment]
> is there a terminology for `Type{T}` now that we don't confuse it with singleton types? If not, do we want to introduce one, eg call it "type selectors"?

Yeah, calling `Type{T}` a singleton sounds a bit confusing since you can write something like `Type{<:Integer}`.  For the new name, "type selectors" sound reasonable to me.

> should I expunge the remaining use of 'singleton' for `Type{T}` from _Conversion and Promotion_?

:+1: from me :)

File: doc/src/manual/types.md
Comment (tpapp) [reply doc/src/manual/types.md:1064]: Yes, I think we should stress that only concrete types can to be singletons.

**Comment by tpapp** [comment]
@tkf: thanks for the review. Yes, sharp field types are another use case --- if you have a small, self-contained example, I would be happy to add it to the PR if that's OK with you.

I included the usage hint because I have seen users on Discourse trying to work with `Type{Type{T}}` etc constructs and other misunderstandings. `Type` is mysterious to a newcomer as is, and each common use case should serve as an example. So I would rather just include them if feasible.

**Comment by tkf** [comment]
Yeah, including example is a good idea.  I think a minimal example is `Some`:

```julia
struct SomeBad{T}
    value::T
end

struct SomeGood{T}
    value::T
end
SomeGood(::Type{T}) where {T} = SomeGood{Type{T}}(T)

sbad = SomeBad(Symbol)
sgood = SomeGood(Symbol)

@assert sbad isa SomeBad{DataType}
@assert sgood isa SomeGood{Type{Symbol}}

getvalue(x) = x.value

@code_warntype getvalue(sbad)  # => Body::DataType
@code_warntype getvalue(sgood)  # => Body::Type{Symbol}
```

**Comment by tpapp** [comment]
Thanks @tkf, I simplified the example somewhat as I did not want to deal with `@code_warntype` in a `jldoctest`. I also did the other changes above.

**Comment by tpapp** [comment]
Friendly bump — if this needs to be improved, I am happy to implement any further suggested changes.

**Comment by StefanKarpinski** [comment]
I've requested review by @JeffBezanson. I'll merge one way or another this week.

File: doc/src/manual/conversion-and-promotion.md
Comment (rfourquet) [inline doc/src/manual/conversion-and-promotion.md:172]: Why not put directly `Type{MyType}` within the `[]` brackets, where `Type{T}` is currently?

File: doc/src/manual/types.md
Comment (rfourquet) [inline doc/src/manual/types.md:1111]: "the `Type{T}` is ..." sounds weird, it reads "the type `T`", where in fact we mean "the `Type{T}` type". I would just remove "the":
```suggestion
For each type `T`, `Type{T}` is an abstract parametric type whose only instance is the
```

File: doc/src/manual/types.md
Comment (rfourquet) [inline doc/src/manual/types.md:1135]:
```suggestion
In other words, [`isa(A, Type{B})`](@ref) is true if and only if `A` and `B` are the same object
```

File: doc/src/manual/types.md
Comment (rfourquet) [inline doc/src/manual/types.md:1138]: Sounds like the sentence is missing a "connection" word, e.g. I would expect "we have:" or something like that at the end, before the example.

File: doc/src/manual/types.md
Comment (rfourquet) [inline doc/src/manual/types.md:1182]:
```suggestion
less precisely, e.g. as [`DataType`](@ref man-declared-types) in the example below where the
```

**Comment by tpapp** [comment]
@rfourquet, thanks for the thorough reading and the detailed suggestions, I applied all of them.

**Review by StefanKarpinski — APPROVED** [review]

**Comment by StefanKarpinski** [comment]
Great clarification. Thank you! This looks ready to merge once CI passes.

File: doc/src/manual/types.md
Comment (dkarrasch) [inline doc/src/manual/types.md:1063]:
```suggestion
Immutable composite types with no fields are called *singletons*. Formally, if
```

File: doc/src/manual/types.md
Comment (rfourquet) [inline doc/src/manual/types.md:1068]: Is there an official convention on whether the footnote sign should go before or after the period? I saw both in the manual.

File: doc/src/manual/types.md
Comment (rfourquet) [inline doc/src/manual/types.md:1065]:
```suggestion
1. `T` is an immutable composite type (i.e. defined with `struct`),
```

**Review by rfourquet — APPROVED** [review]
Yes nice change! Just one more possible change ("ie" -> "i.e.") if you also update this branch for something else.

File: doc/src/manual/types.md
Comment (tpapp) [reply doc/src/manual/types.md:1068]: The [Chicago Manual of Style](https://www.chicagomanualofstyle.org/qanda/data/faq/topics/Punctuation/faq0020.html) suggests that it goes after the period.

Comment (tpapp) [reply doc/src/manual/types.md:1068]: Incidentally, since these questions are in general important for high-quality documentation, I started some discussion about referring a style guide in #36723.

**Comment by tpapp** [comment]
Another friendly bump: I think I addressed all comments.

I am happy to wait for more feedback if necessary, just don't want this PR to be forgotten.

**Comment by tkf** [comment]
It has two approvals and the `merge me` label. I think it's more than enough :)

**Pull request merged**
